### PR TITLE
Add reportUncaughtExceptions option to automatically capture uncaught exceptions

### LINKED ISSUE #99: Automatically capture uncaught exceptions and rejections
Currently, capturing and sending uncaught exceptions and rejections needs to be done manually. Having a helper function that attaches the event handlers would greatly improve the developer experience, and remove the need to use the deprecated Domain API.

### Scope

- Add a helper function to allow uncaught exceptions and rejections to be sent to Raygun automatically (e.g. `raygunClient.reportUncaughtExceptions()`)
  - Use the [uncaughtExceptionMonitor](https://nodejs.org/api/process.html#process_event_uncaughtexceptionmonitor) with a fallback for older versions of Node (Added in: v13.7.0, v12.17.0)
  - Add support for unhandled rejections
- The Domain API is deprecated, so we should remove/replace the [using-domains](https://github.com/MindscapeHQ/raygun4node/tree/master/examples/using-domains) sample app.

### PR DESCRIPTION
**Pull request by Widdershin**
Adds an option when initializing the client `reportUncaughtExceptions`. By default this is false, but when set to true we listen for `uncaughtExceptionMonitor` events and then report them to Raygun.

The catch is that the request has to happen synchronously, which is something that Node doesn't easily facilitate. As such, this PR adds a sync transport that uses `child_process` to perform the request in another process while blocking the parent process.

We had previously discussed having a more complex error monitoring strategy in order to provide support for Node 10, but now that Node 10 is no longer a supported LTS version, I figure it's safe to just provide support for 12 onwards.

I've added a warning message that will log if someone attempts to enable this feature when using an unsupported version.

Additionally, in order to test this with Node 16, I had to upgrade our `tap` version, which also required updating some tets to quash some deprecation messages, but there shouldn't be much risk there.

Fixes #99

### CODE REVIEW

**Review by samuel-holt — APPROVED** [review]
:+1: These changes look good, I was impressed with how you tested the new functionality. I've just pointed out a couple of minor things, but no blockers. 

Did you want to delete/replace the *examples/using-domains* directory as part of this?

File: examples/using-domains/readme.md [hunk 1]
```diff
@@ -2,12 +2,15 @@
 
 A demo on how to use Raygun4Node and a NodeJS domain.
 
+Please note that if you're using Node 12 onwards, we recommend using the `reportUncaughtExceptions` option in place of the `domains` module for reporting fatal errors.
```
Comment (samuel-holt) [inline examples/using-domains/readme.md:5]: :+1:

File: lib/raygun.sync.worker.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
+import fs from "fs";
+
+import * as transport from "./raygun.transport";
+import { SendOptions, SendOptionsWithoutCB } from "./types";
+
+const data = fs.readFileSync(0, "utf-8");
```
[the remark this answers is not in the document]
Comment (Widdershin) [reply lib/raygun.sync.worker.ts:7]: Ah yup good question, I've added a comment making things a bit more explicit.

This does read off of stdin synchronously, and the `0` here is a file descriptor. Conveniently, one of the few things that Windows, macOS and Linux agree on is that the file descriptors for the standard streams are:
0. stdin
1. stdout
2. stderr

File: lib/raygun.sync.worker.ts [hunk 3]
```diff
@@ -3,6 +3,7 @@ import fs from "fs";
 import * as transport from "./raygun.transport";
 import { SendOptions, SendOptionsWithoutCB } from "./types";
 
+// Read stdin synchronously
```
Comment (samuel-holt) [inline lib/raygun.sync.worker.ts:6]: :+1:

File: lib/raygun.sync.transport.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+import type { IncomingMessage, RequestOptions } from "http";
+import {
+  isCallbackWithError,
+  callVariadicCallback,
+  SendOptionsWithoutCB,
+} from "./types";
+
+const debug = require("debug")("raygun");
+
+const API_HOST = "api.raygun.io";
```
[the remark this answers is not in the document]
Comment (Widdershin) [reply lib/raygun.sync.transport.ts]: Nope, good catch, removed 👍

**Review by samuel-holt — APPROVED** [review]
:+1: Thanks for making those changes

File: lib/raygun.sync.transport.ts [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+ *
+ * Copyright (c) 2015 MindscapeHQ
+ * Licensed under the MIT license.
+ */
+
+"use strict";
+
+import { spawnSync } from "child_process";
+
+import type { IncomingMessage, RequestOptions } from "http";
```
[the remark this answers is not in the document]
Comment (Widdershin) [reply lib/raygun.sync.transport.ts]: Now pruned, nice catch.

**Comment by Widdershin** [comment]
Thanks for the detailed review @samuel-holt, good suggestions. And appreciate the fast turnaround for subsequent changes, I hadn't even finished writing my comments before you reviewed it 😅

**Pull request merged**